### ISSUE TITLE
UnorderedMap::value_at should be disabled if is_set is true

### ISSUE THREAD

**Issue by Char-Aznable** [kokkos/kokkos#4555]
In general, the idea of having a foobar value view for set just wrong. A set by definition should not have values associated with the keys.  And this code https://github.com/kokkos/kokkos/blob/ae5fc649ef4b62b48a01123759ed066bff227b43/containers/src/Kokkos_UnorderedMap.hpp#L663-L668 can accidentally return foobar

**Comment by ajpowelsnl** [comment]
Hi - Would you please provide a bit more context to your question.  We don't quite understand the problem.

**Comment by ajpowelsnl** [comment]
Hello -- We would like to help you with this issue, but we need a bit more information to provide a focused response / solution.  When you have a moment, would you please provide more context.

**Comment by Char-Aznable** [comment]
I think a solution would be changing the quoted code above to:
```c++
  template < class I >
  KOKKOS_FORCEINLINE_FUNCTION
  std::enable_if_t<std::is_integral<I>::value && !is_set, Value& >
  value_at(I i) const {
    return m_values[i < capacity() ? i : capacity()];
  }
```
and keep `m_values` member default initialized (zero-sized default view). Without having gone through all the details, I would suggest tighten the loop holes related to `is_set` and `impl_value_type`.

**Comment by Char-Aznable** [comment]
The idea is to have some sort of SFINAE to trigger a compile time error if unntended `value_at()` is called for a `UnorderedMap` which is a set (`is_set` is true)

**Comment by crtrott** [comment]
Hi,

yeah we looked into it and agree basically with your solution. We just need to make it issue a deprecation warning if deprecated code is enabled, and sfinae it away if deprecated code is not disabled. Do you want to issue a PR? If not we can find someone to work on this. 

Christian

**Comment by Char-Aznable** [comment]
Fixed by #4639
